presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/PackageAPI.java

```diff
@@ -0,0 +1,146 @@
+        }
+        catch (MalformedURLException e) {
+            return Response.status(BAD_REQUEST).entity(text("Invalid url")).build();
+        }
+        return Response.status(ACCEPTED).entity(text("Check status to make sure that presto is installed")).build();
+    }
+
+    @POST
+    @Consumes({"text/plain"})
+    @Produces({"text/plain"})
+    @ApiOperation(value = "Upgrade presto")
+    @ApiResponses(value = {
+            @ApiResponse(code = 202, message = "Acknowledged request"),
+            @ApiResponse(code = 400, message = "Bad package format or Invalid url")
+    })
+    public synchronized Response upgrade(@QueryParam("disableDependencyChecking") @DefaultValue("false") boolean disableDependencyChecking,
+            @NotNull String locationToFetchPackage)
+    {
+        try {
+            if (installedPackageType != PackageType.valueOf(getFileExtension(locationToFetchPackage).toUpperCase())) {
+                return Response.status(BAD_REQUEST).entity(text("Provided package format doesn't match the installed package format")).build();
+            }
+        }
+        catch (IllegalArgumentException e) {
+            return Response.status(BAD_REQUEST).entity(text("Invalid package type for presto-server. Package must be .tar.gz or .rpm")).build();
+        }
+        try {
+            final URL url = new URL(locationToFetchPackage);
+            new Thread(new PrestoAsynchronousCommand(new PrestoUpgrader(installedPackageType, url, disableDependencyChecking))).start();
+        }
+        catch (MalformedURLException e) {
+            return Response.status(BAD_REQUEST).entity(text("Invalid url")).build();
+        }
+        return Response.status(ACCEPTED).entity(text("Check status to make sure that presto is upgraded")).build();
+    }
+
+    @DELETE
+    @ApiOperation(value = "Uninstall presto")
+    @ApiResponses(value = {
+            @ApiResponse(code = 202, message = "Acknowledged request")
+    })
+    public synchronized Response uninstall(@QueryParam("disableDependencyChecking") @DefaultValue("false") boolean disableDependencyChecking,
+            @QueryParam("ignoreErrors") @DefaultValue("false") boolean ignoreErrors)
+    {
+        new Thread(new PrestoAsynchronousCommand(new PrestoUninstaller(installedPackageType, disableDependencyChecking, ignoreErrors))).start();
+        return Response.status(ACCEPTED).build();
+    }
+
+    public enum PackageType
+    {
+        RPM, TARBALL;
+    }
+
+    public class PrestoAsynchronousCommand
+            implements Runnable
+    {
+        PrestoCommand command;
+
+        public PrestoAsynchronousCommand(PrestoCommand command)
+        {
+            this.command = command;
+        }
+
+        @Override
+        public void run()
+        {
+            try {
+                command.runCommand();
+            }
+            catch (Exception e) {
```

Comment:
or more to the point you need to log anything weird that happens because otherwise you'll lose the error.